File: scripts/merge-gov.js
#!/usr/bin/env node
/**
 * Merge ministry Excel JSON into an existing src/data bank.
 * Same rules as Install_Prawko.windows.ps1 Merge-GovExcelIntoDataFiles:
 * match on question text; skip when media filename matches or (optional)
 * ffmpeg frames are ≥95% similar. New rows get id suffix -mi when taken.
 *
 *   node scripts/merge-gov.js --gov-dir DIR --out-dir DIR [--ffmpeg PATH] [--media-dir DIR]
 */

const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

const CATEGORIES = ["A", "A1", "A2", "AM", "B", "B1", "C", "C1", "D", "D1", "PT", "T"];
const SIZE = 96;
const THRESHOLD = 0.95;
const EXAM = {
  totalQuestions: 32,
  basicQuestions: 20,
  specialistQuestions: 12,
  maxPoints: 74,
  passThreshold: 68,
  totalTimeSeconds: 1500,
  basicTimeSeconds: 20,
  specialistTimeSeconds: 50,
  basicPoints: [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1],
  specialistPoints: [3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 1, 1],
};

function die(msg) {
  console.error(msg);
  process.exit(1);
}

function parseArgs(argv) {
  const out = { mediaDirs: [] };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const next = () => argv[++i];
    if (a === "--gov-dir" || a === "--GovDir") out.govDir = next();
    else if (a === "--out-dir" || a === "--OutDir") out.outDir = next();
    else if (a === "--ffmpeg") out.ffmpeg = next();
    else if (a === "--ffprobe") out.ffprobe = next();
    else if (a === "--media-dir") out.mediaDirs.push(next());
    else if (a === "--help" || a === "-h") {
      console.log("merge-gov.js --gov-dir DIR --out-dir DIR [--ffmpeg PATH] [--media-dir DIR]");
      process.exit(0);
    } else die("Unknown argument: " + a);
  }
  if (!out.govDir || !out.outDir) die("Required: --gov-dir and --out-dir");
  return out;
}

function readJson(file) {
  return JSON.parse(fs.readFileSync(file, "utf8"));
}

function writeJson(file, data) {
  fs.writeFileSync(file, JSON.stringify(data, null, 2) + "\n", "utf8");
}

function textKey(text) {
  return String(text || "")
    .trim()
    .replace(/\s+/g, " ")
    .toLowerCase();
}

function mediaStem(media) {
  const name = path.basename(String(media || "").trim());
  if (!name) return "";
  return path.parse(name).name.toLowerCase();
}

function uniqueId(desired, existing) {
  let base = String(desired || "").trim() || "mi";
  if (!existing.has(base)) return base;
  let n = 1;
  let id;
  do {
    id = n === 1 ? `${base}-mi` : `${base}-mi${n}`;
    n++;
  } while (existing.has(id));
  return id;
}

function isVideoPath(p, question) {
  if (/\.(mp4|wmv|webm|mov)$/i.test(p)) return true;
  if (/\.(jpg|jpeg|webp|png)$/i.test(p)) return false;
  return String((question && question.mediaType) || "").toLowerCase() === "video";
}

function indexMedia(dirs) {
  const map = new Map();
  for (const dir of dirs) {
    if (!dir || !fs.existsSync(dir)) continue;
    for (const name of fs.readdirSync(dir)) {
      const full = path.join(dir, name);
      try {
        if (!fs.statSync(full).isFile()) continue;
      } catch {
        continue;
      }
      const stem = path.parse(name).name.toLowerCase();
      if (!map.has(stem)) map.set(stem, full);
    }
  }
  return map;
}

function resolveMediaPath(index, question) {
  const stem = mediaStem(question && question.media);
  if (!stem) return "";
  return index.get(stem) || "";
}

function durationSeconds(ffprobe, file) {
  const r = spawnSync(ffprobe, ["-v", "error", "-show_entries", "format=duration", "-of", "csv=p=0", file], {
    encoding: "utf8",
  });
  if (r.status !== 0) return 0;
  const n = parseFloat(String(r.stdout || "").trim());
  return Number.isFinite(n) ? n : 0;
}

function extractFrame(ffmpeg, file, startAt) {
  const r = spawnSync(
    ffmpeg,
    [
      "-y",
      "-hide_banner",
      "-loglevel",
      "error",
      "-ss",
      String(startAt),
      "-i",
      file,
      "-frames:v",
      "1",
      "-vf",
      `scale=${SIZE}:${SIZE}`,
      "-f",
      "rawvideo",
      "-pix_fmt",
      "rgb24",
      "pipe:1",
    ],
    { encoding: "buffer", maxBuffer: SIZE * SIZE * 3 + 1024 }
  );
  if (r.status !== 0 || !r.stdout || !r.stdout.length) return null;
  return Buffer.from(r.stdout);
}

const visualCache = new Map();

function visualSignature(ffmpeg, ffprobe, file, question) {
  const key = file;
  if (visualCache.has(key)) return visualCache.get(key);
  let buf;
  if (isVideoPath(file, question) && ffprobe) {
    const dur = durationSeconds(ffprobe, file);
    const times = dur > 0 ? [0.05, dur / 2, Math.max(0, dur - 0.05)] : [0, 0.5, 1];
    const parts = [];
    for (const t of times) {
      const frame = extractFrame(ffmpeg, file, t);
      if (frame) parts.push(frame);
    }
    buf = parts.length ? Buffer.concat(parts) : extractFrame(ffmpeg, file, 0);
  } else {
    buf = extractFrame(ffmpeg, file, 0);
  }
  visualCache.set(key, buf);
  return buf;
}

function rgbSimilarity(a, b) {
  if (!a || !b) return 0;
  const n = Math.min(a.length, b.length);
  if (!n) return 0;
  let acc = 0;
  for (let i = 0; i < n; i++) acc += Math.abs(a[i] - b[i]);
  return 1 - acc / (n * 255);
}

function visuallySame(ffmpeg, ffprobe, index, left, right) {
  const pathL = resolveMediaPath(index, left);
  const pathR = resolveMediaPath(index, right);
  if (!pathL || !pathR) return false;
  const sigL = visualSignature(ffmpeg, ffprobe, pathL, left);
  const sigR = visualSignature(ffmpeg, ffprobe, pathR, right);
  return rgbSimilarity(sigL, sigR) >= THRESHOLD;
}

function duplicateReason(q, candidates, ffmpeg, ffprobe, index) {
  if (!candidates || !candidates.length) return null;
  const newStem = mediaStem(q.media);
  for (const old of candidates) {
    const oldStem = mediaStem(old.media);
    if (newStem && oldStem && newStem === oldStem) return "filename";
    if (!newStem && !oldStem) return "text-only";
    if (ffmpeg && visuallySame(ffmpeg, ffprobe, index, q, old)) return "visual";
  }
  return null;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const ffmpeg = args.ffmpeg && fs.existsSync(args.ffmpeg) ? args.ffmpeg : "";
  let ffprobe = args.ffprobe || "";
  if (ffmpeg && !ffprobe) {
    const sibling = path.join(path.dirname(ffmpeg), process.platform === "win32" ? "ffprobe.exe" : "ffprobe");
    if (fs.existsSync(sibling)) ffprobe = sibling;
  }
  const mediaIndex = indexMedia(args.mediaDirs);
  if (ffmpeg && !mediaIndex.size) {
    console.log("-> No local media — merge without frame comparison (filename only).");
  } else if (ffmpeg) {
    console.log(`-> Visual media comparison (FFmpeg ${SIZE}×${SIZE}, threshold ${Math.round(THRESHOLD * 100)}%).`);
  } else {
    console.log("-> No FFmpeg — merge without frame comparison (media filename only).");
  }

  let addedTotal = 0;
  const metaCategories = [];
  const useFfmpeg = ffmpeg && mediaIndex.size ? ffmpeg : "";

  for (const cat of CATEGORIES) {
    const src = path.join(args.govDir, `${cat}.json`);
    const dest = path.join(args.outDir, `${cat}.json`);
    if (!fs.existsSync(src)) die(`Missing ${src} — run parse-excel first.`);
    if (!fs.existsSync(dest)) die(`Missing ${dest} — merge needs the repo question bank.`);
    const incoming = readJson(src).questions || [];
    const data = readJson(dest);
    const existingIds = new Set();
    const byText = new Map();
    const list = [];
    for (const q of data.questions || []) {
      existingIds.add(String(q.id).trim());
      const tk = textKey(q.q);
      if (tk) {
        if (!byText.has(tk)) byText.set(tk, []);
        byText.get(tk).push(q);
      }
      list.push(q);
    }

    let added = 0;
    let skippedSame = 0;
    let skippedVisual = 0;
    let idRewritten = 0;
    for (const q of incoming) {
      const tk = textKey(q.q);
      if (!tk) continue;
      const candidates = byText.get(tk) || [];
      const why = duplicateReason(q, candidates, useFfmpeg, ffprobe, mediaIndex);
      if (why === "filename" || why === "text-only") {
        skippedSame++;
        continue;
      }
      if (why === "visual") {
        skippedVisual++;
        continue;
      }
      const newId = uniqueId(q.id, existingIds);
      const toAdd = { ...q };
      if (String(newId) !== String(q.id).trim()) {
        toAdd.id = newId;
        idRewritten++;
      }
      existingIds.add(String(toAdd.id).trim());
      if (!byText.has(tk)) byText.set(tk, []);
      byText.get(tk).push(toAdd);
      list.push(toAdd);
      added++;
    }

    const payload = { category: cat, questions: list };
    writeJson(dest, payload);
    const basic = list.filter((q) => q.type === "basic").length;
    const specialist = list.filter((q) => q.type === "specialist").length;
    metaCategories.push({
      id: cat,
      name: `Kategoria ${cat}`,
      questionCount: list.length,
      basicCount: basic,
      specialistCount: specialist,
    });
    addedTotal += added;
    const skipBits = [];
    if (skippedSame) skipBits.push(`same text+media: ${skippedSame}`);
    if (skippedVisual) skipBits.push(`same text+frames ≥95%: ${skippedVisual}`);
    if (idRewritten) skipBits.push(`new id (number taken): ${idRewritten}`);
    const skipNote = skipBits.length ? `, skipped ${skipBits.join(", ")}` : "";
    console.log(`  ${String(cat).padStart(3)}: +${String(added).padStart(4)} from ministry (total ${list.length}${skipNote})`);
  }

  const metaPath = path.join(args.outDir, "meta.json");
  let exam = EXAM;
  if (fs.existsSync(metaPath)) {
    const old = readJson(metaPath);
    if (old.exam) exam = old.exam;
    for (const cat of metaCategories) {
      const prev = (old.categories || []).find((c) => c.id === cat.id);
      if (prev && prev.name) cat.name = prev.name;
    }
  }
  writeJson(metaPath, { categories: metaCategories, exam });
  console.log(`-> Merge: added ${addedTotal} ministry questions that were not in the repo.`);
}

main();

File: scripts/merge-gov.py
#!/usr/bin/env python3
"""Merge ministry Excel JSON into an existing src/data bank.

Same rules as scripts/merge-gov.ps1:
match on question text; skip when media filename matches or (optional)
ffmpeg frames are ≥95% similar. New rows get id suffix -mi when taken.

  python3 scripts/merge-gov.py --gov-dir DIR --out-dir DIR [--ffmpeg PATH] [--media-dir DIR]
"""

from __future__ import annotations

import argparse
import json
import os
import re
import subprocess
import sys
from pathlib import Path

CATEGORIES = ["A", "A1", "A2", "AM", "B", "B1", "C", "C1", "D", "D1", "PT", "T"]
SIZE = 96
THRESHOLD = 0.95
EXAM = {
    "totalQuestions": 32,
    "basicQuestions": 20,
    "specialistQuestions": 12,
    "maxPoints": 74,
    "passThreshold": 68,
    "totalTimeSeconds": 1500,
    "basicTimeSeconds": 20,
    "specialistTimeSeconds": 50,
    "basicPoints": [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1],
    "specialistPoints": [3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 1, 1],
}


def die(msg: str) -> None:
    print(msg, file=sys.stderr)
    raise SystemExit(1)


def read_json(path: Path):
    return json.loads(path.read_text(encoding="utf-8"))


def write_json(path: Path, data) -> None:
    path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")


def text_key(text) -> str:
    return re.sub(r"\s+", " ", str(text or "").strip()).lower()


def media_stem(media) -> str:
    name = Path(str(media or "").strip()).name
    return Path(name).stem.lower() if name else ""


def unique_id(desired, existing: set) -> str:
    base = str(desired or "").strip() or "mi"
    if base not in existing:
        return base
    n = 1
    while True:
        candidate = f"{base}-mi" if n == 1 else f"{base}-mi{n}"
        if candidate not in existing:
            return candidate
        n += 1


def is_video_path(p: str, question) -> bool:
    if re.search(r"\.(mp4|wmv|webm|mov)$", p, re.I):
        return True
    if re.search(r"\.(jpg|jpeg|webp|png)$", p, re.I):
        return False
    return str((question or {}).get("mediaType") or "").lower() == "video"


def index_media(dirs) -> dict:
    mapping = {}
    for d in dirs:
        if not d:
            continue
        folder = Path(d)
        if not folder.is_dir():
            continue
        for name in os.listdir(folder):
            full = folder / name
            if not full.is_file():
                continue
            stem = Path(name).stem.lower()
            mapping.setdefault(stem, str(full))
    return mapping


def resolve_media_path(index: dict, question) -> str:
    stem = media_stem((question or {}).get("media"))
    return index.get(stem, "") if stem else ""


def duration_seconds(ffprobe: str, file: str) -> float:
    r = subprocess.run(
        [ffprobe, "-v", "error", "-show_entries", "format=duration", "-of", "csv=p=0", file],
        capture_output=True,
        text=True,
        check=False,
    )
    if r.returncode != 0:
        return 0.0
    try:
        return float((r.stdout or "").strip())
    except ValueError:
        return 0.0


def extract_frame(ffmpeg: str, file: str, start_at: float):
    r = subprocess.run(
        [
            ffmpeg, "-y", "-hide_banner", "-loglevel", "error",
            "-ss", str(start_at), "-i", file, "-frames:v", "1",
            "-vf", f"scale={SIZE}:{SIZE}", "-f", "rawvideo", "-pix_fmt", "rgb24", "pipe:1",
        ],
        capture_output=True,
        check=False,
    )
    if r.returncode != 0 or not r.stdout:
        return None
    return r.stdout


_visual_cache = {}


def visual_signature(ffmpeg: str, ffprobe: str, file: str, question):
    if file in _visual_cache:
        return _visual_cache[file]
    if is_video_path(file, question) and ffprobe:
        dur = duration_seconds(ffprobe, file)
        times = [0.05, dur / 2, max(0, dur - 0.05)] if dur > 0 else [0, 0.5, 1]
        parts = []
        for t in times:
            frame = extract_frame(ffmpeg, file, t)
            if frame:
                parts.append(frame)
        buf = b"".join(parts) if parts else extract_frame(ffmpeg, file, 0)
    else:
        buf = extract_frame(ffmpeg, file, 0)
    _visual_cache[file] = buf
    return buf


def rgb_similarity(a, b) -> float:
    if not a or not b:
        return 0.0
    n = min(len(a), len(b))
    if not n:
        return 0.0
    acc = sum(abs(a[i] - b[i]) for i in range(n))
    return 1 - acc / (n * 255)


def visually_same(ffmpeg, ffprobe, index, left, right) -> bool:
    path_l = resolve_media_path(index, left)
    path_r = resolve_media_path(index, right)
    if not path_l or not path_r:
        return False
    return rgb_similarity(
        visual_signature(ffmpeg, ffprobe, path_l, left),
        visual_signature(ffmpeg, ffprobe, path_r, right),
    ) >= THRESHOLD


def duplicate_reason(q, candidates, ffmpeg, ffprobe, index):
    if not candidates:
        return None
    new_stem = media_stem(q.get("media"))
    for old in candidates:
        old_stem = media_stem(old.get("media"))
        if new_stem and old_stem and new_stem == old_stem:
            return "filename"
        if not new_stem and not old_stem:
            return "text-only"
        if ffmpeg and visually_same(ffmpeg, ffprobe, index, q, old):
            return "visual"
    return None


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--gov-dir", required=True)
    ap.add_argument("--out-dir", required=True)
    ap.add_argument("--ffmpeg", default="")
    ap.add_argument("--ffprobe", default="")
    ap.add_argument("--media-dir", action="append", default=[])
    args = ap.parse_args()

    ffmpeg = args.ffmpeg if args.ffmpeg and Path(args.ffmpeg).exists() else ""
    ffprobe = args.ffprobe
    if ffmpeg and not ffprobe:
        sibling = Path(ffmpeg).with_name("ffprobe")
        if sibling.exists():
            ffprobe = str(sibling)
    media_index = index_media(args.media_dir)
    if ffmpeg and not media_index:
        print("-> No local media — merge without frame comparison (filename only).")
        ffmpeg = ""
    elif ffmpeg:
        print(f"-> Visual media comparison (FFmpeg {SIZE}×{SIZE}, threshold {round(THRESHOLD * 100)}%).")
    else:
        print("-> No FFmpeg — merge without frame comparison (media filename only).")

    added_total = 0
    meta_categories = []
    gov_dir = Path(args.gov_dir)
    out_dir = Path(args.out_dir)

    for cat in CATEGORIES:
        src = gov_dir / f"{cat}.json"
        dest = out_dir / f"{cat}.json"
        if not src.is_file():
            die(f"Missing {src} — run parse-excel first.")
        if not dest.is_file():
            die(f"Missing {dest} — merge needs the repo question bank.")
        incoming = read_json(src).get("questions") or []
        data = read_json(dest)
        existing_ids = set()
        by_text = {}
        lst = []
        for q in data.get("questions") or []:
            existing_ids.add(str(q.get("id", "")).strip())
            tk = text_key(q.get("q"))
            if tk:
                by_text.setdefault(tk, []).append(q)
            lst.append(q)

        added = skipped_same = skipped_visual = id_rewritten = 0
        for q in incoming:
            tk = text_key(q.get("q"))
            if not tk:
                continue
            why = duplicate_reason(q, by_text.get(tk) or [], ffmpeg, ffprobe, media_index)
            if why in ("filename", "text-only"):
                skipped_same += 1
                continue
            if why == "visual":
                skipped_visual += 1
                continue
            new_id = unique_id(q.get("id"), existing_ids)
            to_add = dict(q)
            if str(new_id) != str(q.get("id", "")).strip():
                to_add["id"] = new_id
                id_rewritten += 1
            existing_ids.add(str(to_add["id"]).strip())
            by_text.setdefault(tk, []).append(to_add)
            lst.append(to_add)
            added += 1

        write_json(dest, {"category": cat, "questions": lst})
        basic = sum(1 for q in lst if q.get("type") == "basic")
        specialist = sum(1 for q in lst if q.get("type") == "specialist")
        meta_categories.append({
            "id": cat,
            "name": f"Kategoria {cat}",
            "questionCount": len(lst),
            "basicCount": basic,
            "specialistCount": specialist,
        })
        added_total += added
        skip_bits = []
        if skipped_same:
            skip_bits.append(f"same text+media: {skipped_same}")
        if skipped_visual:
            skip_bits.append(f"same text+frames ≥95%: {skipped_visual}")
        if id_rewritten:
            skip_bits.append(f"new id (number taken): {id_rewritten}")
        skip_note = f", skipped {', '.join(skip_bits)}" if skip_bits else ""
        print(f"  {cat:>3}: +{added:4} from ministry (total {len(lst)}{skip_note})")

    meta_path = out_dir / "meta.json"
    exam = EXAM
    if meta_path.is_file():
        old = read_json(meta_path)
        if old.get("exam"):
            exam = old["exam"]
        for cat in meta_categories:
            prev = next((c for c in (old.get("categories") or []) if c.get("id") == cat["id"]), None)
            if prev and prev.get("name"):
                cat["name"] = prev["name"]
    write_json(meta_path, {"categories": meta_categories, "exam": exam})
    print(f"-> Merge: added {added_total} ministry questions that were not in the repo.")


if __name__ == "__main__":
    main()
